### PR TITLE
Refactorisation of MonadTimers & monad transformer instances

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

File: io-classes/src/Control/Monad/Class/MonadTimer.hs [hunk 1]
```diff
@@ -60,6 +63,79 @@ defaultRegisterDelay d = do
...
+    maxDelay = microsecondsAsIntToDiffTime maxBound
+
+    -- The timeout thread, it might be killed by an async exception. In this
+    -- case the `cancel` action is responsible for updating state state of the
+    -- timeout (held in `v`).
+    go :: TVar m (Maybe (Timeout m))
+       -> TVar m TimeoutState
+       -> Time
+       -> Time
+       -> m ()
```
Comment (coot) [inline io-classes/src/Control/Monad/Class/MonadTimer.hs]: experiment with a slightly differnt design.

File: si-timers/src/Control/Monad/Class/MonadTimer/NonStandard.hs [hunk 2]
```diff
@@ -24,12 +24,11 @@ module Control.Monad.Class.MonadTimer.NonStandard
   , TimeoutState (..)
   ) where
 
-import qualified Control.Concurrent.STM.TVar as STM
+import qualified Control.Concurrent.Class.MonadSTM as STM
```
Comment (coot) [inline si-timers/src/Control/Monad/Class/MonadTimer/NonStandard.hs:27]: Revert this change.

File: io-classes/src/Control/Monad/Class/MonadThrow.hs [hunk 3]
```diff
@@ -339,3 +346,104 @@ instance MonadMask m => MonadMask (ExceptT e m) where
...
+-- StateT instances
+--
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadThrow (StateT s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadCatch (StateT s m) where
+  catch (StateT m) f = StateT $ \s -> catch (m s) (\e -> runStateT (f e) s)
```
Comment (coot) [inline io-classes/src/Control/Monad/Class/MonadThrow.hs:428]: compare with `exceptions` package

File: io-classes/src/Control/Monad/Class/MonadSTM/Internal.hs [hunk 4]
```diff
@@ -1303,21 +1300,24 @@ catchSTM = MonadThrow.catch
 data Trans where
     Cont   :: Trans
     Reader :: Trans
```
Comment (coot) [inline io-classes/src/Control/Monad/Class/MonadSTM/Internal.hs:1302]: unfinshed buissness!

**Comment by coot** [comment]
notes after a conversation with @dcoutts:

* [ ] `MonadTimerSI` -> `IOSIm`, remove `MonadTimeout`
* [ ] Make `io-sim` and `si-timers` Time independent.
* [ ] check which instances  `excpetions` package provides
* [ ] `io-classes-mtl` (leave `ReaderT` instances in `io-classes`)

File: io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+-- StateT instances
+--
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadThrow (StateT s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadCatch (StateT s m) where
+  catch (StateT m) f = StateT $ \s -> catch (m s) (\e -> runStateT (f e) s)
```
Comment (coot) [inline io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs:133]: This is compatible with what exception package does; it's using [`liftCatch`](https://hackage.haskell.org/package/transformers-0.6.0.4/docs/src/Control.Monad.Trans.State.Strict.html#liftCatch).

File: io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+-- These all follow the @exceptions@ package to the letter
+--
+
+instance MonadCatch m => MonadThrow (ExceptT e m) where
+  throwIO = lift . throwIO
+
+instance MonadCatch m => MonadCatch (ExceptT e m) where
+  catch (ExceptT m) f = ExceptT $ catch m (runExceptT . f)
+
+  generalBracket acquire release use = ExceptT $ do
```
Comment (coot) [inline io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs:24]: The same as https://hackage.haskell.org/package/exceptions-0.10.7/docs/src/Control.Monad.Catch.html#local-6989586621679054762

File: io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadThrow (WriterT w m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadCatch (WriterT w m) where
+  catch (WriterT m) f = WriterT $ catch m (runWriterT . f)
+
+  generalBracket acquire release use = WriterT $ fmap f $
```
Comment (coot) [inline io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs:66]: The same as https://hackage.haskell.org/package/exceptions-0.10.7/docs/src/Control.Monad.Catch.html#local-6989586621679055342

File: io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadThrow (RWST r w s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadCatch (RWST r w s m) where
+  catch (RWST m) f = RWST $ \r s -> catch (m r s) (\e -> runRWST (f e) r s)
+
+  -- | general bracket ignores the state produced by the release callback
+  generalBracket acquire release use = RWST $ \r s ->
```
Comment (coot) [inline io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs:108]: The same as https://hackage.haskell.org/package/exceptions-0.10.7/docs/src/Control.Monad.Catch.html#local-6989586621679055102

File: io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadThrow (StateT s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadCatch (StateT s m) where
+  catch (StateT m) f = StateT $ \s -> catch (m s) (\e -> runStateT (f e) s)
+
+  -- | general bracket ignores the state produced by the release callback
+  generalBracket acquire release use = StateT $ \s -> fmap f $
```
Comment (coot) [inline io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs:150]: The same as https://hackage.haskell.org/package/exceptions-0.10.7/docs/src/Control.Monad.Catch.html#local-6989586621679055622

File: io-classes/src/Control/Monad/Class/MonadThrow.hs (hunk 3 above)
Comment (coot) [reply io-classes/src/Control/Monad/Class/MonadThrow.hs]: https://hackage.haskell.org/package/transformers-0.6.0.4/docs/src/Control.Monad.Trans.State.Lazy.html#liftCatch

File: io-classes/src/Control/Monad/Class/MonadTimer.hs [hunk 10]
```diff
@@ -60,6 +63,79 @@ defaultRegisterDelay d = do
...
+    go tv v c u | u `diffTime` c >= maxDelay = do
+      t <- newTimeout maxDelay
+      _ <- atomically $ swapTVar tv $! Just t
```
[the remark this answers is not in the document]
Comment (coot) [reply io-classes/src/Control/Monad/Class/MonadTimer.hs]: I think it was, but later by calling `awaitTimeout`, but this was refactored; we no longer store `t` in a `TVar`.

File: io-classes/src/Control/Monad/Class/MonadTimer.hs [hunk 11]
```diff
@@ -60,6 +63,79 @@ defaultRegisterDelay d = do
...
+                         -> m (STM m TimeoutState, m ())
+
+registerDelayCancellable d | d <= maxDelay = do
+    t <- newTimeout d
+    return (readTimeout t, cancelTimeout t)
+  where
+    maxDelay :: DiffTime
+    maxDelay = microsecondsAsIntToDiffTime maxBound
+
+registerDelayCancellable d = do
```
[the remark this answers is not in the document]
Comment (coot) [reply io-classes/src/Control/Monad/Class/MonadTimer.hs]: `io-classes` do not protect for `32` bit overflow, the primitives in `si-timers` do.

File: io-classes/src/Control/Monad/Class/MonadTimer.hs [hunk 12]
```diff
@@ -28,132 +24,34 @@ import           Control.Monad.Trans (lift)
...
 
-import           Control.Monad.Class.MonadFork
-import           Control.Monad.Class.MonadTime
 import           Control.Monad.Class.MonadTimer.NonStandard
 
 import qualified System.Timeout as IO
 
 class Monad m => MonadDelay m where
-  threadDelay :: DiffTime -> m ()
+  threadDelay :: Int -> m ()
```
[the remark this answers is not in the document]
Comment (coot) [reply io-classes/src/Control/Monad/Class/MonadTimer.hs:22]: Probably, but I am not sure if I do that.

File: io-classes/src/Control/Monad/Class/MonadSTM/Internal.hs [hunk 13]
```diff
@@ -1303,21 +1300,24 @@ catchSTM = MonadThrow.catch
...
+--
 newtype WrappedSTM (t :: Trans) r (m :: Type -> Type) a = WrappedSTM { runWrappedSTM :: STM m a }
 
 deriving instance MonadSTM m => Functor     (WrappedSTM t r m)
 deriving instance MonadSTM m => Applicative (WrappedSTM t r m)
 deriving instance MonadSTM m => Monad       (WrappedSTM t r m)
-deriving instance MonadSTM m => Alternative (WrappedSTM t r m)
-deriving instance MonadSTM m => MonadPlus   (WrappedSTM t r m)
+-- deriving instance (MonadSTM m, Alternative m) => Alternative (WrappedSTM t r m)
+-- deriving instance (MonadSTM m, MonadPlus m)   => MonadPlus   (WrappedSTM t r m)
```
[the remark this answers is not in the document]
Comment (coot) [reply io-classes/src/Control/Monad/Class/MonadSTM/Internal.hs]: This was removed later.

File: io-classes/src/Control/Monad/Class/MonadSTM/Internal.hs [hunk 14]
```diff
@@ -1303,21 +1300,24 @@ catchSTM = MonadThrow.catch
...
     Cont   :: Trans
     Reader :: Trans
-    Writer :: Trans
-    State  :: Trans
-    Except :: Trans
-    RWS    :: Trans
+    -- Writer :: Trans
+    -- State  :: Trans
+    -- Except :: Trans
+    -- RWS    :: Trans
```
[the remark this answers is not in the document]
Comment (coot) [reply io-classes/src/Control/Monad/Class/MonadSTM/Internal.hs]: I was removed in a later patch.

File: io-sim/src/Control/Monad/IOSim/Internal.hs [hunk 15]
```diff
@@ -380,7 +379,7 @@ schedule !thread@Thread{
     NewTimeout d k | d < 0 ->
       {-# SCC "schedule.NewTimeout.1" #-} do
       let !t       = NegativeTimeout nextTmid
-          !expiry  = microsecondsAsIntToDiffTime d `addTime` time
+          !expiry  = d `addTime` time
```
[the remark this answers is not in the document]
Comment (coot) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:382]: Yes, that's rather unfortunate in this PR.  I switched to time in `Word64` in `io-sim` to switch back to it after introducing `MonadTimer.SI`

**Review by dcoutts — APPROVED** [review]
LGTM. Phew!

File: si-timers/src/Control/Monad/Class/MonadTimer/SI.hs [hunk 16]
```diff
@@ -35,20 +35,29 @@ import           Data.Time.Clock (diffTimeToPicoseconds)
 
 
 
+-- | Convert 'DiffTime' in seconds to microseconds represented by an 'Int'.
+--
+-- Note that on 32bit systems it can only represent 2^31 seconds, which is only
```
Comment (bolt12) [inline si-timers/src/Control/Monad/Class/MonadTimer/SI.hs:43]: Isn't it 2^31 - 1?

File: io-classes-mtl/io-classes-mtl.cabal [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+name:               io-classes-mtl
+version:            0.1.0.0
+synopsis:           Experimental MTL instances for io-classes
+license:            Apache-2.0
+license-files:
+  LICENSE
+  NOTICE
+author:             Duncan Coutts, Marcin Szamotulski
+maintainer:         Marcin Szamotulski coot@coot.me
+copyright:          2022 Input Output (Hong Kong) Ltd.
```
Comment (bolt12) [inline io-classes-mtl/io-classes-mtl.cabal:11]: 2023?

File: si-timers/src/Control/Monad/Class/MonadTimer/SI.hs [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+          _            -> return ()
+
+
+-- | Like 'GHC.Conc.registerDelay' but safe on 32-bit systems.  When the delay
+-- is larger than what `Int` can represent it will fork a thread which will
+-- write to the returned 'TVar' once the delay has passed.  When the delay is
+-- small enough it will use the `MonadTimer`'s `registerDelay` (e.g. for `IO`
+-- monad it will use the `GHC`'s `GHC.Conc.registerDelay`).
+--
+-- TODO: 'timeout' not safe on 32-bit systems.
```
[the remark this answers is not in the document]
Comment (coot) [reply si-timers/src/Control/Monad/Class/MonadTimer/SI.hs:235]: That's not trivial, so I prefer to leave it for now.
https://github.com/input-output-hk/io-sim/issues/84

File: si-timers/src/Control/Monad/Class/MonadTimer/SI.hs [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+    v <- atomically $ newTVar False
+    _ <- forkIO $ go v c (d `addTime` c)
+    return v
+  where
+    maxDelay :: DiffTime
+    maxDelay = microsecondsAsIntToDiffTime maxBound
+
+    go :: TVar m Bool -> Time -> Time -> m ()
+    go v c u | u `diffTime` c >= maxDelay = do
+      _ <- newTimeout maxBound >>= atomically . awaitTimeout
```
[the remark this answers is not in the document]
Comment (coot) [reply si-timers/src/Control/Monad/Class/MonadTimer/SI.hs:144]: but then we'd need to add `MonadDelay` constraint.

File: io-sim/io-sim.cabal [hunk 20]
```diff
@@ -7,7 +7,7 @@ license:             Apache-2.0
 license-files:
   LICENSE
   NOTICE
-copyright:           2019-2020 Input Output (Hong Kong) Ltd.
+copyright:           2019-2022 Input Output (Hong Kong) Ltd.
```
[the remark this answers is not in the document]
Comment (coot) [reply io-sim/io-sim.cabal:10]: Yes, fixed.

File: io-classes/io-classes.cabal [hunk 21]
```diff
@@ -24,7 +24,19 @@ flag asserts
...
 
+common warnings
+    ghc-options: -Wall
+                 -Wcompat
+                 -Wincomplete-uni-patterns
+                 -Wincomplete-record-updates
+                 -Wpartial-fields
+                 -Widentities
+                 -Wunused-packages
+                 -Wno-redundant-constraints
```
[the remark this answers is not in the document]
Comment (coot) [reply io-classes/io-classes.cabal:35]: What are you referring to?

[the remark this answers is not in the document]
Comment (bolt12) [reply io-classes/io-classes.cabal:35]: That's what the commit message says

**Review by bolt12 — APPROVED** [review]
